File: projects/rpn2tex/migrations_archive/go-feature-by-feature-1/ast.go
package rpn2tex

// Expr is the interface for all AST expression nodes.
type Expr interface {
	Position() (line int, column int)
	Precedence() int
}

// NumberNode represents a numeric literal in the AST.
type NumberNode struct {
	Line   int
	Column int
	Value  string // String representation preserves formatting
}

// Position returns the line and column of the node.
func (n *NumberNode) Position() (int, int) {
	return n.Line, n.Column
}

// Precedence returns the precedence level of the node.
// Numbers have the highest precedence (3) and never need parentheses.
func (n *NumberNode) Precedence() int {
	return 3
}

// BinaryOpNode represents a binary operation in the AST.
type BinaryOpNode struct {
	Line     int
	Column   int
	Operator string // Operator symbol ("+", "-", "*", "/")
	Left     Expr   // Left operand expression
	Right    Expr   // Right operand expression
}

// Position returns the line and column of the node.
func (b *BinaryOpNode) Position() (int, int) {
	return b.Line, b.Column
}

// Precedence returns the precedence level of the operator.
// Addition and subtraction have precedence 1 (lowest).
// Multiplication and division have precedence 2 (higher).
func (b *BinaryOpNode) Precedence() int {
	switch b.Operator {
	case "+", "-":
		return 1
	case "*", "/":
		return 2
	default:
		return 0
	}
}

File: projects/rpn2tex/migrations_archive/rust-feature-by-feature-1/src/ast.rs
//! AST node definitions for rpn2tex parser.
//!
//! This module defines the Abstract Syntax Tree (AST) nodes used to represent
//! parsed RPN expressions.

/// Expression node types.
///
/// Represents the structure of mathematical expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    /// Numeric literal
    Number(Number),
    /// Binary operation (e.g., addition, subtraction)
    BinaryOp(BinaryOp),
}

/// Numeric literal node.
///
/// Represents integer and decimal numbers in expressions.
///
/// # Examples
///
/// ```
/// use rpn2tex::ast::Number;
///
/// let num = Number::new("42".to_string(), 1, 1);
/// assert_eq!(num.value, "42");
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Number {
    /// The string representation of the number
    pub value: String,
    /// Line number (1-based) where node appears
    pub line: usize,
    /// Column number (1-based) where node starts
    pub column: usize,
}

impl Number {
    /// Create a new Number node.
    ///
    /// # Examples
    ///
    /// ```
    /// use rpn2tex::ast::Number;
    ///
    /// let num = Number::new("3.14".to_string(), 1, 1);
    /// ```
    #[must_use]
    pub fn new(value: String, line: usize, column: usize) -> Self {
        Self {
            value,
            line,
            column,
        }
    }
}

/// Binary operation node.
///
/// Represents operations with two operands (e.g., addition, subtraction).
///
/// # Examples
///
/// ```
/// use rpn2tex::ast::{BinaryOp, Expr, Number};
///
/// let left = Expr::Number(Number::new("5".to_string(), 1, 1));
/// let right = Expr::Number(Number::new("3".to_string(), 1, 3));
/// let add = BinaryOp::new("+".to_string(), left, right, 1, 5);
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BinaryOp {
    /// The operator string (e.g., "+", "-", "*", "/")
    pub operator: String,
    /// Left operand
    pub left: Box<Expr>,
    /// Right operand
    pub right: Box<Expr>,
    /// Line number (1-based) where node appears
    pub line: usize,
    /// Column number (1-based) where node starts
    pub column: usize,
}

impl BinaryOp {
    /// Create a new BinaryOp node.
    ///
    /// # Examples
    ///
    /// ```
    /// use rpn2tex::ast::{BinaryOp, Expr, Number};
    ///
    /// let left = Expr::Number(Number::new("5".to_string(), 1, 1));
    /// let right = Expr::Number(Number::new("3".to_string(), 1, 3));
    /// let add = BinaryOp::new("+".to_string(), left, right, 1, 5);
    /// ```
    #[must_use]
    pub fn new(operator: String, left: Expr, right: Expr, line: usize, column: usize) -> Self {
        Self {
            operator,
            left: Box::new(left),
            right: Box::new(right),
            line,
            column,
        }
    }
}
